File: backend/src/main/java/com/flowlink/mock/MockRuntime.java
package com.flowlink.mock;

import com.flowlink.mock.MockHttp.FiredCallback;
import com.flowlink.mock.MockHttp.MockRequest;
import com.flowlink.mock.MockHttp.MockResponse;
import com.flowlink.mock.MockSpec.MockCond;
import com.flowlink.mock.MockSpec.MockRoute;
import com.flowlink.mock.MockSpec.MockRule;
import org.springframework.stereotype.Component;

import java.nio.charset.Charset;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * CUSTOM mock 서버의 순수 런타임 — 라우트 매칭·조건 평가·템플릿 렌더.
 * 상태가 없어 단위 테스트 대상. (콜백 "발사"는 {@link MockCallbackDispatcher}가 담당)
 *
 * <p>템플릿 토큰(응답 body·헤더 값·콜백 url/body):
 * {@code {{path.x}} {{query.x}} {{header.x}} {{body.x}} {{body}} {{method}} {{uuid}} {{seq}} {{now}}}
 * — 미해석 토큰은 빈 문자열. (워크플로 바인딩 {@code {{ key@node }}} 와는 다른 문맥)
 */
@Component
public class MockRuntime {

    private static final Pattern TOKEN = Pattern.compile("\\{\\{\\s*(.+?)\\s*}}");
    public static final int MAX_DELAY_MS = 10_000;
    public static final int MAX_CALLBACK_DELAY_MS = 60_000;

    /** 매칭 결과 — 규칙·경로 파라미터. */
    public record Match(MockRule rule, Map<String, String> pathParams) {
    }

    /** 정의 순서대로 method+경로 첫 매칭 라우트, 그 안에서 조건 만족 첫 규칙. */
    public Optional<Match> match(List<MockRoute> routes, MockRequest req) {
        for (MockRoute route : routes) {
            Map<String, String> params = matchPath(route.path(), req.path());
            if (params == null) {
                continue;
            }
            String m = route.method() == null ? "ANY" : route.method().toUpperCase(Locale.ROOT);
            if (!m.equals("ANY") && !m.equals(req.method())) {
                continue;
            }
            for (MockRule rule : route.rulesOrEmpty()) {
                if (conditionsPass(rule.whenOrEmpty(), req, params)) {
                    return Optional.of(new Match(rule, params));
                }
            }
            // 경로는 맞지만 규칙 무매칭 — 다음 라우트로 넘기지 않고 404 (같은 경로 중복 정의 혼란 방지)
            return Optional.empty();
        }
        return Optional.empty();
    }

    /** 규칙 → 실제 응답(바이트) + 지연 + 콜백 명세. seq 는 서버별 증가 카운터 공급자에서 받은 값. */
    public MockResponse render(MockRule rule, MockRequest req, Map<String, String> pathParams, long seq) {
        Charset cs = MockHttp.charsetOf(rule.charset());
        String body = template(rule.body() == null ? "" : rule.body(), req, pathParams, seq);
        Map<String, String> headers = new LinkedHashMap<>();
        for (MockSpec.KV kv : rule.headers() == null ? List.<MockSpec.KV>of() : rule.headers()) {
            if (kv.key() != null && !kv.key().isBlank()) {
                headers.put(kv.key(), template(kv.value() == null ? "" : kv.value(), req, pathParams, seq));
            }
        }
        int delay = rule.delayMs() == null ? 0 : Math.max(0, Math.min(rule.delayMs(), MAX_DELAY_MS));
        FiredCallback cb = null;
        MockSpec.MockCallback c = rule.callback();
        if (c != null) {
            String url = template(c.url() == null ? "" : c.url(), req, pathParams, seq).trim();
            if (!url.isEmpty()) {
                cb = new FiredCallback(
                        c.afterMs() == null ? 0 : Math.max(0, Math.min(c.afterMs(), MAX_CALLBACK_DELAY_MS)),
                        url,
                        c.method() == null || c.method().isBlank() ? "POST" : c.method().toUpperCase(Locale.ROOT),
                        MockHttp.contentTypeHeader(
                                c.contentType() == null || c.contentType().isBlank() ? "urlencoded" : c.contentType(),
                                java.nio.charset.StandardCharsets.UTF_8),
                        template(c.body() == null ? "" : c.body(), req, pathParams, seq),
                        Boolean.TRUE.equals(c.retryUntilOk()));
            }
        }
        int status = rule.status() == null ? 200 : rule.status();
        return new MockResponse(status, MockHttp.contentTypeHeader(rule.contentType(), cs), headers,
                body.getBytes(cs), delay, cb);
    }

    // ---------- 경로 매칭 ----------

    /** /users/{id} 패턴 매칭. 매칭되면 경로 파라미터 맵(빈 맵 가능), 아니면 null. */
    static Map<String, String> matchPath(String pattern, String actual) {
        if (pattern == null || pattern.isBlank()) {
            return null;
        }
        String[] ps = normalize(pattern).split("/");
        String[] as = normalize(actual).split("/");
        if (ps.length != as.length) {
            return null;
        }
        Map<String, String> params = new LinkedHashMap<>();
        for (int i = 0; i < ps.length; i++) {
            String p = ps[i];
            if (p.length() >= 2 && p.startsWith("{") && p.endsWith("}")) {
                if (as[i].isEmpty()) {
                    return null;
                }
                params.put(p.substring(1, p.length() - 1), as[i]);
            } else if (!p.equals(as[i])) {
                return null;
            }
        }
        return params;
    }

    private static String normalize(String path) {
        String p = path == null ? "" : path.trim();
        if (!p.startsWith("/")) {
            p = "/" + p;
        }
        if (p.length() > 1 && p.endsWith("/")) {
            p = p.substring(0, p.length() - 1);
        }
        return p;
    }

    // ---------- 조건 ----------

    static boolean conditionsPass(List<MockCond> conds, MockRequest req, Map<String, String> pathParams) {
        for (MockCond c : conds) {
            String actual = valueOf(c.source(), c.key(), req, pathParams);
            String op = c.op() == null ? "eq" : c.op().toLowerCase(Locale.ROOT);
            boolean pass = switch (op) {
                case "eq" -> actual != null && actual.equals(c.value() == null ? "" : c.value());
                case "ne" -> actual == null || !actual.equals(c.value() == null ? "" : c.value());
                case "exists" -> actual != null && !actual.isEmpty();
                case "contains" -> actual != null && actual.contains(c.value() == null ? "" : c.value());
                default -> false;
            };
            if (!pass) {
                return false;
            }
        }
        return true;
    }

    private static String valueOf(String source, String key, MockRequest req, Map<String, String> pathParams) {
        if (key == null) {
            return null;
        }
        String src = source == null ? "" : source.toLowerCase(Locale.ROOT);
        return switch (src) {
            case "query" -> req.query().get(key);
            case "header" -> req.headers().get(key.toLowerCase(Locale.ROOT));
            case "body" -> req.bodyFields().get(key);
            case "path" -> pathParams.get(key);
            default -> null;
        };
    }

    // ---------- 템플릿 ----------

    String template(String text, MockRequest req, Map<String, String> pathParams, long seq) {
        if (text == null || text.isEmpty() || !text.contains("{{")) {
            return text == null ? "" : text;
        }
        Matcher m = TOKEN.matcher(text);
        StringBuilder sb = new StringBuilder();
        while (m.find()) {
            m.appendReplacement(sb, Matcher.quoteReplacement(resolve(m.group(1), req, pathParams, seq)));
        }
        m.appendTail(sb);
        return sb.toString();
    }

    private String resolve(String token, MockRequest req, Map<String, String> pathParams, long seq) {
        String t = token.trim();
        String v = switch (t) {
            case "uuid" -> UUID.randomUUID().toString();
            case "seq" -> String.valueOf(seq);
            case "now" -> java.time.Instant.now().toString();
            case "method" -> req.method();
            case "body" -> req.bodyText() == null ? "" : req.bodyText();
            default -> null;
        };
        if (v != null) {
            return v;
        }
        int dot = t.indexOf('.');
        if (dot > 0 && dot < t.length() - 1) {
            String src = t.substring(0, dot);
            String key = t.substring(dot + 1);
            String got = valueOf(src, key, req, pathParams);
            return got == null ? "" : got;
        }
        return ""; // 미해석 토큰 — 빈 문자열
    }
}

File: backend/src/main/kotlin/com/flowlink/mock/MockRuntime.kt
package com.flowlink.mock

import com.fasterxml.jackson.databind.ObjectMapper
import com.flowlink.mock.MockHttp.FiredCallback
import com.flowlink.mock.MockHttp.MockRequest
import com.flowlink.mock.MockHttp.MockResponse
import com.flowlink.mock.MockSpec.MockCond
import com.flowlink.mock.MockSpec.MockRoute
import com.flowlink.mock.MockSpec.MockRule
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.util.LinkedHashMap
import java.util.Locale
import java.util.Optional

/**
 * 사용자 정의 mock 라우트 실행기(순수 — 저장소/HTTP 의존 없음).
 *
 * 매칭: 정의 순서대로 method+경로 패턴(`/users/{id}`) 첫 매칭 라우트 → 그 안에서 조건(AND) 만족 첫 규칙.
 * 렌더: 규칙의 본문/헤더/setState/콜백 템플릿을 [MockTemplate] 문맥(요청·경로·seq·상태·시크릿)으로 해석.
 * 템플릿 문법은 [MockTemplate] — `{{ x@body }}`(칩) 와 `{{body.x}}`(dot) 둘 다, `{{ 이름@secret }}` 포함.
 * 미해석 토큰은 빈 문자열(워크플로 바인딩 `{{ key@node }}` 와는 다른 문맥).
 */
@Component
class MockRuntime {

    /** 매칭 결과 — [req] 는 라우트 코덱(prepare)까지 적용된 요청(조건·템플릿이 본 그대로). */
    data class Match(val rule: MockRule, val pathParams: Map<String, String>, val req: MockRequest)

    /** 응답 후 코덱 훅 — 렌더된 본문 + (기록 가능한) 헤더 + 규칙 contentType → 최종 본문. */
    fun interface ResponseCodec {
        fun apply(body: String, headers: MutableMap<String, String>, contentType: String?): String
    }

    /**
     * 정의 순서대로 method+경로 첫 매칭 라우트, 그 안에서 조건 만족 첫 규칙.
     * [prepare] 는 경로/메서드가 맞은 라우트에 대해 조건 평가 **전** 요청을 바꿀 기회(요청 코덱 — 본문 디코딩).
     */
    fun match(
        routes: List<MockRoute>, req: MockRequest, state: Map<String, String> = emptyMap(), hits: Map<String, Int> = emptyMap(),
        prepare: ((MockRoute, MockRequest) -> MockRequest)? = null
    ): Optional<Match> {
        for (route in routes) {
            val params = matchPath(route.path, req.path) ?: continue
            val m = if (route.method == null) "ANY" else route.method.uppercase(Locale.ROOT)
            if (m != "ANY" && m != req.method) {
                continue
            }
            val r = if (prepare == null) req else prepare(route, req)
            for (rule in route.rulesOrEmpty()) {
                // 순차 응답: repeat 소진(처음 N회 매칭 후)이면 이 규칙은 건너뛰고 다음 규칙으로 폴스루
                if (rule.repeat != null && rule.id != null && (hits[rule.id] ?: 0) >= rule.repeat) {
                    continue
                }
                if (conditionsPass(rule.whenOrEmpty(), r, params, state)) {
                    return Optional.of(Match(rule, params, r))
                }
            }
            // 경로는 맞지만 규칙 무매칭 — 다음 라우트로 넘기지 않고 404 (같은 경로 중복 정의 혼란 방지)
            return Optional.empty()
        }
        return Optional.empty()
    }

    /**
     * 규칙 → 실제 응답(바이트) + 지연 + 콜백 명세. seq 는 서버별 증가 카운터 공급자에서 받은 값.
     * [responseCodec] 은 템플릿 렌더가 끝난 본문 전체에 적용(응답 코덱 — 문자셋 인코딩 직전, 헤더 기록 가능). 콜백 본문에는 미적용.
     * [secrets] 는 `{{ 이름@secret }}` 해석용(게이트웨이가 Mock 시크릿 환경으로 조회), [json] 은 본문 점 경로 토큰용.
     */
    fun render(
        rule: MockRule, req: MockRequest, pathParams: Map<String, String>, seq: Long, state: Map<String, String> = emptyMap(),
        responseCodec: ResponseCodec? = null, secrets: Map<String, String> = emptyMap(), json: ObjectMapper? = null
    ): MockResponse {
        val ctx = MockContext(req = req, pathParams = pathParams, seq = seq, state = state, secrets = secrets, json = json)
        val cs = MockHttp.charsetOf(rule.charset)
        val rendered = MockTemplate.render(rule.body ?: "", ctx)
        val headers = LinkedHashMap<String, String>()
        for (kv in rule.headers ?: emptyList()) {
            if (kv.key != null && kv.key.isNotBlank()) {
                headers[kv.key] = MockTemplate.render(kv.value ?: "", ctx)
            }
        }
        val body = if (responseCodec == null) rendered else responseCodec.apply(rendered, headers, rule.contentType)
        // 상태 있는 목: setState 의 값을 템플릿 해석해 서버 상태에 반영할 맵으로(게이트웨이가 적용).
        // op=incr/decr 은 현재 상태값을 숫자로 누산(재고·잔액 원장 시뮬레이션). set(기본)은 대입.
        val newState = LinkedHashMap<String, String>()
        for (kv in rule.setState ?: emptyList()) {
            if (kv.key == null || kv.key.isBlank()) continue
            val v = MockTemplate.render(kv.value ?: "", ctx)
            newState[kv.key] = when (kv.op?.lowercase(Locale.ROOT)) {
                "incr" -> ((state[kv.key]?.toLongOrNull() ?: 0L) + (v.toLongOrNull() ?: 1L)).toString()
                "decr" -> ((state[kv.key]?.toLongOrNull() ?: 0L) - (v.toLongOrNull() ?: 1L)).toString()
                else -> v
            }
        }
        val delay = if (rule.delayMs == null) 0 else Math.max(0, Math.min(rule.delayMs, MAX_DELAY_MS))
        var cb: FiredCallback? = null
        val c = rule.callback
        if (c != null) {
            val url = MockTemplate.render(c.url ?: "", ctx).trim()
            if (url.isNotEmpty()) {
                cb = FiredCallback(
                    if (c.afterMs == null) 0 else Math.max(0, Math.min(c.afterMs, MAX_CALLBACK_DELAY_MS)),
                    url,
                    if (c.method == null || c.method.isBlank()) "POST" else c.method.uppercase(Locale.ROOT),
                    MockHttp.contentTypeHeader(
                        if (c.contentType == null || c.contentType.isBlank()) "urlencoded" else c.contentType,
                        StandardCharsets.UTF_8
                    ),
                    MockTemplate.render(c.body ?: "", ctx),
                    c.retryUntilOk == true
                )
            }
        }
        val status = rule.status ?: 200
        return MockResponse(
            status, MockHttp.contentTypeHeader(rule.contentType, cs), headers,
            body.toByteArray(cs), delay, cb, newState
        )
    }

    // ---------- 템플릿(호환 편의) ----------

    fun template(text: String?, req: MockRequest, pathParams: Map<String, String>, seq: Long, state: Map<String, String> = emptyMap()): String =
        MockTemplate.render(text, MockContext(req = req, pathParams = pathParams, seq = seq, state = state))

    companion object {
        const val MAX_DELAY_MS = 10_000
        const val MAX_CALLBACK_DELAY_MS = 60_000

        // ---------- 경로 매칭 ----------

        /** /users/{id} 패턴 매칭. 매칭되면 경로 파라미터 맵(빈 맵 가능), 아니면 null. */
        @JvmStatic
        fun matchPath(pattern: String?, actual: String?): Map<String, String>? {
            if (pattern == null || pattern.isBlank()) {
                return null
            }
            val ps = normalize(pattern).split("/")
            val ac = normalize(actual).split("/")
            if (ps.size != ac.size) {
                return null
            }
            val params = LinkedHashMap<String, String>()
            for (i in ps.indices) {
                val p = ps[i]
                if (p.length >= 2 && p.startsWith("{") && p.endsWith("}")) {
                    if (ac[i].isEmpty()) {
                        return null
                    }
                    params[p.substring(1, p.length - 1)] = ac[i]
                } else if (p != ac[i]) {
                    return null
                }
            }
            return params
        }

        private fun normalize(path: String?): String {
            var p = path?.trim() ?: ""
            if (!p.startsWith("/")) {
                p = "/$p"
            }
            if (p.length > 1 && p.endsWith("/")) {
                p = p.substring(0, p.length - 1)
            }
            return p
        }

        // ---------- 조건 ----------

        @JvmStatic
        fun conditionsPass(conds: List<MockCond>, req: MockRequest, pathParams: Map<String, String>, state: Map<String, String> = emptyMap()): Boolean {
            val ctx = MockContext(req = req, pathParams = pathParams, state = state)
            for (c in conds) {
                val actual = MockTemplate.valueOf(c.source, c.key, ctx)
                val op = if (c.op == null) "eq" else c.op.lowercase(Locale.ROOT)
                val cv = c.value ?: ""
                val pass = when (op) {
                    "eq" -> actual != null && actual == cv
                    "ne" -> actual == null || actual != cv
                    "exists" -> actual != null && actual.isNotEmpty()
                    "contains" -> actual != null && actual.contains(cv)
                    "startswith" -> actual != null && actual.startsWith(cv)
                    "endswith" -> actual != null && actual.endsWith(cv)
                    "gt", "gte", "lt", "lte" -> numCompare(actual, cv, op)
                    "regex" -> actual != null && try { Regex(cv).containsMatchIn(actual) } catch (e: Exception) { false }
                    else -> false
                }
                if (!pass) {
                    return false
                }
            }
            return true
        }

        /** 숫자 비교(gt/gte/lt/lte) — 양변이 숫자가 아니면 false. */
        private fun numCompare(actual: String?, value: String, op: String): Boolean {
            val a = actual?.toDoubleOrNull() ?: return false
            val b = value.toDoubleOrNull() ?: return false
            return when (op) { "gt" -> a > b; "gte" -> a >= b; "lt" -> a < b; "lte" -> a <= b; else -> false }
        }
    }
}
